### PR TITLE
Fallback instead of failing for cases where are not able to set the socket options

### PR DESCRIPTION
**Pull request by yashykt**
Fallback instead of failing for cases where are not able to set the socket options. Taking @ncteisen 's suggestion on failing softly if we are not able to set the socket options. Now, we will clean up the buffer list and fallback on writing without timestamps.

### CODE REVIEW

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE                                       FILE SIZE
 ++++++++++++++ GROWING                         ++++++++++++++

 -------------- SHRINKING                       --------------
  [ = ]       0 [None]                             -144  -0.0%
  -0.7%     -64 src/core/lib/iomgr/tcp_posix.cc     -64  -0.7%
      -3.8%     -51 tcp_flush                           -51  -3.8%
      -5.7%     -13 [Unmapped]                          -13  -5.7%

  -0.0%     -64 TOTAL                              -208  -0.0%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
Objective-C binary sizes
*****************STATIC******************
  New size                      Old size
 2,020,437      Total (>)      2,020,410

 No significant differences in binary sizes

***************FRAMEWORKS****************
  New size                      Old size
11,174,074      Total (<)     11,174,080

 No significant differences in binary sizes


```

File: src/core/lib/iomgr/tcp_posix.cc [hunk 1]
```diff
@@ -852,13 +849,17 @@ static bool tcp_flush(grpc_tcp* tcp, grpc_error** error) {
     msg.msg_iov = iov;
     msg.msg_iovlen = iov_size;
     msg.msg_flags = 0;
+    bool tried_sending_message = false;
     if (tcp->outgoing_buffer_arg != nullptr) {
-      if (!tcp_write_with_timestamps(tcp, &msg, sending_length, &sent_length,
-                                     error)) {
+      if (!tcp_write_with_timestamps(tcp, &msg, sending_length, &sent_length)) {
```
Comment (ncteisen) [inline src/core/lib/iomgr/tcp_posix.cc]: should we cache this and not try again if the platform does not support?

**Comment by grpc-testing** [comment]
```
Corrupt JSON data (indicates timeout or crash): 
    bm_call_create.BM_IsolatedFilter_ClientChannelFilter_NoOp_.counters.new: 10
    bm_call_create.BM_IsolatedFilter_ClientChannelFilter_NoOp_.counters.old: 10


[microbenchmarks] No significant performance differences
```

File: src/core/lib/iomgr/tcp_posix.cc (hunk 1 above)
Comment (yashykt) [reply src/core/lib/iomgr/tcp_posix.cc]: done

Comment (yashykt) [reply src/core/lib/iomgr/tcp_posix.cc]: I made it per fd for now. In the future if we are sure that is not per fd rather for the entire platform, then we can have a process wide flag

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE                                       FILE SIZE
 ++++++++++++++ GROWING                         ++++++++++++++

 -------------- SHRINKING                       --------------
  [ = ]       0 [None]                             -200  -0.0%
  -0.3%     -32 src/core/lib/iomgr/tcp_posix.cc     -32  -0.3%
      -1.7%     -22 tcp_flush                           -22  -1.7%
      -4.4%     -10 [Unmapped]                          -10  -4.4%

  -0.0%     -32 TOTAL                              -232  -0.0%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

File: src/core/lib/iomgr/tcp_posix.cc [hunk 2]
```diff
@@ -852,13 +850,18 @@ static bool tcp_flush(grpc_tcp* tcp, grpc_error** error) {
...
     if (tcp->outgoing_buffer_arg != nullptr) {
-      if (!tcp_write_with_timestamps(tcp, &msg, sending_length, &sent_length,
-                                     error)) {
-        tcp_shutdown_buffer_list(tcp);
-        return true; /* something went wrong with timestamps */
+      if (!tcp->ts_capable ||
+          !tcp_write_with_timestamps(tcp, &msg, sending_length, &sent_length)) {
+        /* We could not set socket options to collect Fathom timestamps.
+         * Fallback on writing without timestamps. */
+        tcp->ts_capable = false tcp_shutdown_buffer_list(tcp);
```
[the remark this answers is not in the document]
Comment (soheilhy) [reply src/core/lib/iomgr/tcp_posix.cc]: Ah never mind. The new commit fixed this.

**Comment by grpc-testing** [comment]
```
Corrupt JSON data (indicates timeout or crash): 
    bm_call_create.BM_IsolatedFilter_ClientChannelFilter_NoOp_.counters.new: 10
    bm_call_create.BM_IsolatedFilter_ClientChannelFilter_NoOp_.counters.old: 10


[microbenchmarks] No significant performance differences
```

File: src/core/lib/iomgr/tcp_posix.cc (hunk 2 above)
[the remark this answers is not in the document]
Comment (yashykt) [reply src/core/lib/iomgr/tcp_posix.cc]: hahahah :D

**Comment by grpc-testing** [comment]
```
****************************************************************

libgrpc.so

     VM SIZE                                       FILE SIZE
 ++++++++++++++ GROWING                         ++++++++++++++

 -------------- SHRINKING                       --------------
  [ = ]       0 [None]                             -168  -0.0%
  -0.3%     -32 src/core/lib/iomgr/tcp_posix.cc     -32  -0.3%
      -1.7%     -22 tcp_flush                           -22  -1.7%
      -4.4%     -10 [Unmapped]                          -10  -4.4%

  -0.0%     -32 TOTAL                              -200  -0.0%


****************************************************************

libgrpc++.so

     VM SIZE        FILE SIZE
 ++++++++++++++  ++++++++++++++

  [ = ]       0        0  [ = ]



```

**Comment by grpc-testing** [comment]
```
[trickle] No significant performance differences
```

**Comment by grpc-testing** [comment]
```
[microbenchmarks] No significant performance differences
```

**Comment by yashykt** [comment]
Known issues #17526, #15464

**Comment by yashykt** [comment]
Thanks for reviewing!

**Pull request merged**